File: src/solar/util/math/MathUtil.ts
const PI: number = 3.14159
const E: number = 2.71828
const CELLWIDTH: number = 128.0
const CLIFFHEIGHT: number = 128.0
const UNIT_FACING: number = 270.0
const RADTODEG: number = 180.0 / PI
const DEGTORAD: number = PI / 180.0
/** @noSelf **/
export default class MathUtil {


    /**
     * 2点之间的距离
     * @param x1
     * @param y1
     * @param x2
     * @param y2
     */
    static distanceBetweenPoints(x1: number, y1: number, x2: number, y2: number): number {
        let dx = x2 - x1
        let dy = y2 - y1
        return SquareRoot(dx * dx + dy * dy)
    }

    /**
     * 极坐标位移 获得指定方向的x,y
     * @param x1
     * @param y1
     * @param dist
     * @param angle
     */
    static polarProjection(x1: number, y1: number, dist: number, angle: number): { x: number, y: number } {
        let x = x1 + dist * Cos(angle * DEGTORAD)
        let y = y1 + dist * Sin(angle * DEGTORAD)
        return {x: x, y: y}
    }

    /**
     * 与用户交互的 使用角度便于理解和编辑
     * @param x1
     * @param y1
     * @param x2
     * @param y2
     */
    static angleBetweenCoords(x1: number, y1: number, x2: number, y2: number): number {
        return RADTODEG * Atan2(y2 - y1, x2 - x1)
    }

    /**
     * 大多数底层计算函数都是用这个 弧度
     * @param x1
     * @param y1
     * @param x2
     * @param y2
     */
    static radianBetweenCoords(x1: number, y1: number, x2: number, y2: number): number {
        return Atan2(y2 - y1, x2 - x1)
    }

    /**
     * 弧度转角度
     * @param radian
     */
    static radian2angle(radian: number) {
        return RADTODEG * radian
    }

    /**
     * 角度转弧度
     * @param angle
     */
    static angle2radian(angle: number) {
        return DEGTORAD * angle
    }

    /**
     * 判断角度是否在另一角度的背面
     * @param jd0 背部单位到前单位角度
     * @param jd1 前单位面向角度
     */
    static isBackAngle(jd0: number, jd1: number): boolean {
        return CosBJ(jd1 - jd0) >= 0
    }

    /**
     * 统计 一个对象的所有属性的值
     * @param data
     */
    static sum(data: { [key: string]: number }): number {
        if (!data) {
            return 0;
        }
        let count = 0;
        for (let k in data) {
            count = count + data[k];
        }
        return count;
    }


    /**
     * 获取最小值
     */
    static min(...values: number[]): number {
        return Math.min(...values)
    }

    /**
     * 获取最大值
     * @param values
     */
    static max(...values: number[]): number {
        return Math.max(...values)
    }

    /**
     * 将一个数值限制到一个大小范围
     * 低于这个范围最小值 则返回范围最小值
     * 大于这个范围最大值 则返回范围最大值
     * 大于等于范围最小值小于等于范围最大值则返回原值
     * @param value
     * @param min
     * @param max
     */
    static clamp(value: number, min: number, max: number): number {
        if (value > max) {
            return max;
        } else if (value < min) {
            return min;
        }
        return value;
    }

    /**
     * 计算护甲减少伤害比例
     * @param armor 护甲
     * @param ardf 护甲减少伤害因子
     *      负护甲             正护甲
     * 返回  -0.99999... 到 0.999999....
     */
    static armorReduction(armor: number, ardf = ArmorReducesDamageFactor): number {
        if (armor == 0) {
            return 0;
        } else if (armor < 0) {
            armor = -armor;
            return -(armor / (armor + (1 / ardf)));
        }
        return armor / (armor + (1 / ardf));
    }


}

File: scripts/solar/util/math/MathUtil.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__SourceMapTraceBack = ____lualib.__TS__SourceMapTraceBack
__TS__SourceMapTraceBack(debug.getinfo(1).short_src, {["6"] = 1,["7"] = 2,["8"] = 3,["9"] = 4,["10"] = 5,["11"] = 6,["12"] = 7,["13"] = 9,["14"] = 9,["15"] = 9,["17"] = 9,["18"] = 19,["19"] = 20,["20"] = 21,["21"] = 22,["22"] = 19,["23"] = 32,["24"] = 33,["25"] = 34,["26"] = 35,["27"] = 32,["28"] = 45,["29"] = 46,["30"] = 45,["31"] = 56,["32"] = 57,["33"] = 56,["34"] = 64,["35"] = 65,["36"] = 64,["37"] = 72,["38"] = 73,["39"] = 72,["40"] = 81,["41"] = 82,["42"] = 81,["43"] = 89,["44"] = 90,["45"] = 91,["47"] = 93,["48"] = 94,["49"] = 95,["51"] = 97,["52"] = 89,["53"] = 104,["54"] = 105,["55"] = 104,["56"] = 112,["57"] = 113,["58"] = 112,["59"] = 125,["60"] = 126,["61"] = 127,["62"] = 128,["63"] = 129,["65"] = 131,["66"] = 125,["67"] = 141,["68"] = 141,["69"] = 141,["71"] = 142,["72"] = 143,["73"] = 144,["74"] = 145,["75"] = 146,["77"] = 148,["78"] = 141,["79"] = 9});
local ____exports = {}
local PI = 3.14159
local E = 2.71828
local CELLWIDTH = 128
local CLIFFHEIGHT = 128
local UNIT_FACING = 270
local RADTODEG = 180 / PI
local DEGTORAD = PI / 180
____exports.default = __TS__Class()
local MathUtil = ____exports.default
MathUtil.name = "MathUtil"
function MathUtil.prototype.____constructor(self)
end
function MathUtil.distanceBetweenPoints(x1, y1, x2, y2)
    local dx = x2 - x1
    local dy = y2 - y1
    return SquareRoot(dx * dx + dy * dy)
end
function MathUtil.polarProjection(x1, y1, dist, angle)
    local x = x1 + dist * Cos(angle * DEGTORAD)
    local y = y1 + dist * Sin(angle * DEGTORAD)
    return {x = x, y = y}
end
function MathUtil.angleBetweenCoords(x1, y1, x2, y2)
    return RADTODEG * Atan2(y2 - y1, x2 - x1)
end
function MathUtil.radianBetweenCoords(x1, y1, x2, y2)
    return Atan2(y2 - y1, x2 - x1)
end
function MathUtil.radian2angle(radian)
    return RADTODEG * radian
end
function MathUtil.angle2radian(angle)
    return DEGTORAD * angle
end
function MathUtil.isBackAngle(jd0, jd1)
    return CosBJ(jd1 - jd0) >= 0
end
function MathUtil.sum(data)
    if not data then
        return 0
    end
    local count = 0
    for k in pairs(data) do
        count = count + data[k]
    end
    return count
end
function MathUtil.min(...)
    return math.min(...)
end
function MathUtil.max(...)
    return math.max(...)
end
function MathUtil.clamp(value, min, max)
    if value > max then
        return max
    elseif value < min then
        return min
    end
    return value
end
function MathUtil.armorReduction(armor, ardf)
    if ardf == nil then
        ardf = ArmorReducesDamageFactor
    end
    if armor == 0 then
        return 0
    elseif armor < 0 then
        armor = -armor
        return -(armor / (armor + 1 / ardf))
    end
    return armor / (armor + 1 / ardf)
end
____exports.default = MathUtil
return ____exports
